File: leet-code/206. Reverse Linked List/ReverseLinkedList206_3.java
public class ReverseLinkedList206_3 {
    public ListNode reverseList(ListNode head) {
        return reverse(head, null);
    }

    private ListNode reverse(ListNode cur, ListNode prev) {
        if (cur == null) {
            return prev;
        }

        ListNode next = cur.next;
        cur.next = prev;
        return reverse(next, cur);
    }
}

File: leet-code/206. Reverse Linked List/ReverseLinkedList206_3.kt
package com.example

class ReverseLinkedList206_3 {
    fun reverseList(head: ListNode?): ListNode? {
        var cur = head
        var reversedHead: ListNode? = null
        while (cur != null) {
            val next = cur.next
            cur.next = reversedHead
            reversedHead = cur
            cur = next
        }
        return reversedHead
    }
}
